pkg/inject/server.go

```diff
@@ -88,61 +87,66 @@ func (whsvr *WebhookServer) mutate(ar *v1beta1.AdmissionReview) *v1beta1.Admissi
-    containerMode, err := getAnnotation(&pod.ObjectMeta, annotationContainerModeKey)
-    containerName, err := getAnnotation(&pod.ObjectMeta, annotationContainerNameKey)
-
-
-    var sidecarConfig *PatchConfig
-    switch injectType {
-    case "secretless":
-        secretlessConfigMapName, err := getAnnotation(&pod.ObjectMeta, annotationSecretlessConfigKey)
-        if err != nil {
-            return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s", pod.Name, req.Namespace, err.Error()))
-        }
-
-        conjurConnConfigMapName, _ := getAnnotation(&pod.ObjectMeta, annotationConjurConnConfigKey)
-        conjurAuthConfigMapName, _ := getAnnotation(&pod.ObjectMeta, annotationConjurAuthConfigKey)
-
-        sidecarConfig = generateSecretlessSidecarConfig(
-            secretlessConfigMapName,
-            conjurConnConfigMapName,
-            conjurAuthConfigMapName)
-        break;
-    case "authenticator":
-        conjurAuthConfigMapName, err := getAnnotation(&pod.ObjectMeta, annotationConjurAuthConfigKey)
-        if err != nil {
-            return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s", pod.Name, req.Namespace, err.Error()))
-        }
-
-        conjurConnConfigMapName, err := getAnnotation(&pod.ObjectMeta, annotationConjurConnConfigKey)
-        if err != nil {
-            return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s", pod.Name, req.Namespace, err.Error()))
-        }
-
-        switch containerMode {
-        case "sidecar", "init", "":
-            break;
-        default:
-            return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s value (%s) not supported", pod.Name, req.Namespace, annotationContainerModeKey, containerMode))
-        }
-
-        sidecarConfig = generateAuthenticatorSidecarConfig(AuthenticatorSidecarConfig{
-            conjurConnConfigMapName: conjurConnConfigMapName,
-            conjurAuthConfigMapName: conjurAuthConfigMapName,
-            containerMode:           containerMode,
-            containerName:           containerName,
-        })
-        break;
-    default:
-        errMsg := fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to invalid inject type annotation value = %s", pod.Name, req.Namespace, injectType)
-        glog.Infof(errMsg)
-        return &v1beta1.AdmissionResponse{
-            Result: &metav1.Status{
-                Message: errMsg,
-            },
-        }
-    }
+	ag := &annotationGetter{metadata: &pod.ObjectMeta}
+	injectType := ag.getAnnotation(annotationInjectTypeKey)
+	containerMode := ag.getAnnotation(annotationContainerModeKey)
+	containerName := ag.getAnnotation(annotationContainerNameKey)
+	if ag.err != nil {
+		return failWithResponse(fmt.Sprintf("Mutation failed for pod %s, in namespace %s, due to %s", pod.Name, req.Namespace, ag.err.Error()))
```

Comment:
This is a nice pattern :sparkles:. Errors are indeed values!

Reply:
Though it had a side effect of making all annotations required, which wasn't the previous behaviour or in line with the docs, so I've reverted it.